src/datasets/table.py

```diff
@@ -1798,6 +1799,79 @@ def wrapper(array, *args, **kwargs):
     return wrapper
 
 
+def array_concat(arrays: List[pa.Array]):
+    """Improved version of pa.concat_arrays
+
+    It supports concatenating pa.ExtensionArray objects by concatenating the underlying storages.
+
+    Args:
+        arrays (List[pa.Array]): List of arrays to contatenate
+
+    Raises:
+        pa.ArrowInvalid: if the arrow array concatenation fails
+        ValueError: if the list of arrays is empty
+        TypeError: if the arrays to be concatenated have different types
+
+    Returns:
+        array (:obj:`pyarrow.Array`): the concatenated array
+    """
+    arrays = list(arrays)
+    array_types = set(array.type for array in arrays)
+
+    if not array_types:
+        raise ValueError("Couldn't concatenate empty list of arrays")
+    if len(array_types) > 1:
+        array_types = list(array_types)
+        raise TypeError(f"Couldn't concatenate arrays with different types {array_types[0]} and {array_types[1]}")
+
+    def _offsets_concat(offsets):
+        offset = offsets[0]
+        concatenated_offsets = offset
+        for offset in offsets[1:]:
+            offset = pc.subtract(offset, offset[0])
+            offset = pc.add(offset[1:], concatenated_offsets[-1])
+            concatenated_offsets = pa.concat_arrays([concatenated_offsets, offset])
+        return concatenated_offsets
+
+    # breakpoint_ = arrays = [pa.array([{"a": 21, "b": [[1, 2], [3]]}]), pa.array([{"a": 100, "b": [[1], None]}])]
+
+    def _concat_arrays(arrays):
```

Comment:
Isn't it possible to use `pa.concat_arrays` directly here ? It can be used for regular arrays, and also for the storage of extension arrays

Reply:
I don't think it is. PyArrow extension arrays can be nested inside a list/struct, so we need to call `_concat_arrays` (recursively) to extract their storage arrays before we concatenate them (`pa.concat_arrays` would fail in this scenario).

Another option would be to do the following:
1. run a procedure that extracts the storage arrays
2. concatenate the arrays
3. re-cast to the original type

But this seems more expensive than the current solution.

Reply:
Ok I see :) sounds good then